src/movement/ThrusterController.cpp

```diff
@@ -0,0 +1,187 @@
+#include "movement/ThrusterController.hpp"
+namespace rs
+{
+    /**
+     * Constructor.
+     *
+     * @param thrusters The number of thrusters controlled by the controller.
+     * @param max_speed The maximum normalized speed in the forward and reverse
+     *        directions that should be allowed by the controller.
+     * @param delay_ms Delay specified in milliseconds to continue sending
+     *        zeroing signal to continue thruster operation.
+     * @param port The serial port to be used for communicating with the
+     *        thrusters. This port may be changed during runtime.
+     */
+    ThrusterController::ThrusterController(const int thrusters,
+                                           const double max_speed,
+                                           Serial *port,
+                                           const double delay_ms) :
+            _port(port),
+            _thrusters(_thrusters),
+            _max_speed(_max_speed),
+            _delay_ms(delay_ms)
+    {
+        if (delay_ms < 185.00)
+        {
+            ROS_ERROR("Thruster millisecond delay is too low.");
+        }
+    }
+
+    /**
+     * Deconstructor.
+     *
+     * @brief This function sends zeroing commands to the maestro on class
+     *        destruction.
+     */
+    ThrusterController::~ThrusterController()
+    {
+        if (_port != nullptr)
+        {
+            vector<double> zero(_thrusters, 0.0);
+            set(zero);
+        }
+    }
+
+    /**
+     * Sets the serial port used to communicate with the thrusters.
+     *
+     * @param port A pointer to a configured and initialized serial port to be
+     *        used for communicating with the Maestro.
+     *
+     * @return Zero upon success and -1 upon failure.
+     */
+    int ThrusterController::setPort(Serial *port)
+    {
+        _port = port;
+        if (_port != nullptr)
+        {
+            uint8_t detect_byte = 0xAA;
+            if (port->Write(&detect_byte, 1) != 1)
+            {
+                ROS_ERROR("Serial port failed to write baud-detection bit.");
+                return -1;
+            }
+            return 0;
+        }
+        ROS_ERROR("Serial port pointer supplied is not a valid pointer.");
+        return -1;
+    }
+
+    /**
+     * Sets the speed of a particular thruster.
+     *
+     * @param speeds The normalized speeds to set all thrusters to.
+     *
+     * @return 0 on success and -1 on failure.
+     */
+    int ThrusterController::set(vector<double> &speeds)
```

Comment:
Simply taking in a vector of speeds will not work if we decide to disable a single thruster. Also, thrusters messages from up above are not necessarily in the order of ports. I recommend taking in a second vector of ports/channels that describe which motor the speed goes to.